README.md
# GIBBS
Gliders’ in-browser basic soccer monitor, a web based player for RoboCup log files

    rcg2html converter

    version: 2.0
    Authors: Edward Moore, Oliver Obst
    Date:    05/05/13
    License: GPL3

***********************************************

1. OVERVIEW
===================


This is a set of scripts to help display rcg logfiles from RoboCup
soccer simulation league in a web browser. The idea of this HTML5
logplayer is to avoid the use of flash, since compiling the original
flash converter by Thilo Girmann requires a proprietary library that
is no longer available. 

If you find these scripts useful, e.g., for showing matches on your
web page, please do cite our team description report in your papers
and/or team descriptions. This report is an extended version of the
original Gliders2012 team report, the reference is as follows:


Gliders2012: Development and Competition Results
Edward Moore, Oliver Obst, Mikhail Prokopenko, Peter Wang, Jason Held
arXiv:1211.3882
http://arxiv.org/abs/1211.3882


It's perfectly OK to modify and re-distribute modified versions of
this tool. If you do, please keep the message above intact. See also
the file LICENSE for details.

If you find bugs or have suggestions for improvements, or would like
to contribute fixes or improvements, please send us an email
(oliverobst@gmail.com). We will try to incorporate your patches and
suggestions in a next version.


2. SETUP
===================

There is a number of steps to follow initial installation of these
scripts. Files can be installed on a web server by simply copying (run
in a web browser).

The tool to convert log files into the *.replay files required by our
log player needs to be compiled first in order to be run.


2.1 Compiling the Converter

In the "converter" subdirectory, there is a very simple Makefile that 
should do the job. You will need the rcgbase installed that comes with
the server. Just do:

    make

2.2 Converting rcg[.gz] files 

To convert .rcg(.gz) files into a .replay (converted game replay) is 
now very simple:

  The rcg2replay program will convert .rcg or gzipped rcg files into text, 
  e.g.,

  rcg2replay 201203051705-TeamEI_8-vs-HELIOS_base_1.rcg.gz > 201203051705-TeamEI_8-vs-HELIOS_base_1.replay


Repeat for all games you want to show on your web page.


2.3 Installing the scripts on the web server

 - Copy all generated .replay files into some directory on your web
   server (for minimal changes, retain the directory structure coming
   with these scripts). 

 - play.php is the main file that starts up the logplayer when invoked 
   using a browser.  Copy this file into the desired directory.

   This file should be modified to include all the .replay files. If
   you copied them into a different subdirectory, make sure to include
   the correct path. See also online examples at
   http://www.oliverobst.eu/research/robotics-gliders2012-simulation-league-robocup-team/robocup-2012

 - Copy the scripts directory and the *.css files. 
   play.php includes 3 javascript files, and 2 CSS style sheets. All 3
   javascript files are expected in a subdirectory "scripts". This
   subdirectory should be in the same directory as play.php. The 2 CSS
   files are expected to be in the same directory as play.php. 

 - Copy the images directory.
   The style sheet jquery-ui.css expects some icons and graphics in
   the "images" subdirectory (which again should be in the same
   directory as play.php).


3. PLAYING BACK
===================

The example web page (play.php) gives you a drop-down menu to select
available games. To play back a game, it needs to be transferred to
the client side. Currently, the typical size of a replay file is
4-5MB.  This transfer can take time and may make the browser seem like
it has frozen. A notice to say that the file is being loaded would
probably be nice - feel free to add this or any other feature to the
php script. 


4. DEVELOPMENT
===================

Further development of the rcg2html tool may change the format of
replay files to reduce size.


ChangeLog
2013-05-05  Oliver Obst  <oliverobst@gmail.com>

	* scripts/player.js:
	Changed the script to work with the new replay format. This one also displays 
	player direction, neck direction, and stamina (via body color).

	* play.php: 
	- loading the .replay files from the replay subdirectory
	- removed one of the javascript dependencies

	* converter/rcg2replay.cpp: 
	This converter replaces the previous one, and is able to work 
	directly with .rcg files, i.e., no intermediate step is required.
	A very simple Makefile was also added - it will need the rcgbase
	library installed on your system.

	* logplay.css:
	fixed background image path


scripts/player.js
/*
 *	Robocup TXT player
 *	Copyright: Edward Moore, Oliver Obst, Mikhail Prokopenko
 *	Date: July 2012
 *	
 *	A javascript program that will read a file on the server that has been processed by rcg2txt (provided with RCSS Server)
 *	& ParseTXT (Custom formatting program).
*/

var DEBUG		= true;
var OFFSET_W            = 26;
var OFFSET_H            = 29;
var FIELD_W		= 748; /* 800; */
var FIELD_H	 	= 472; /* 514; */
var SCALING	 	= 7;
var PLAYER_SIZE	= 7;
var L_COLOUR	= "#FFFF00";
var LG_COLOUR 	= "#8B008B";
var R_COLOUR	= "#FF0000";
var RG_COLOUR	= "#7FFF00";
var B_COLOUR	= "#FFFFFF";
var CBG			= 10;
var cycles 		= 1;
var count 		= 1;
var L_GOALS		= 0;
var R_GOALS		= 0;
var Lteam		= "";
var Rteam		= "";

var prevState = 0;

var context;
var content;
var play;

function toRad(angle)
{
    if (angle<0) angle += 360.0;
    if (angle>360) angle -= 360.0;

    return Math.PI * angle / 180.0;
}

function toGreyRGB(stamina)
{
    if (stamina.slice(0,1) == "#")
    {
	stamina = stamina.slice(1);
    }
    var max = 8000.0;
    var step = max/256.0;
    val = Math.min(parseFloat(stamina), max);
    val = Math.min(Math.floor((max - val)/step), 255);
    grey = val.toString(16);
    grey = grey.length == 1 ? "0" + grey : grey;
    return "#" + grey + grey + grey;
}

function Pause(){ clearInterval(play); play = false; }
function PlayOne(){
	context.clearRect(0,0,FIELD_W+2*OFFSET_W,FIELD_H+2*OFFSET_H);

	for(var i = 0; i < 22+2; i++)
		MakeMove(content[count++]);
}

function RPlayOne(){
	context.clearRect(0,0,FIELD_W+2*OFFSET_W,FIELD_H+2*OFFSET_H);

	for(var i=0; i < 22+2; i++)
		MakeMove(content[--count]);
}

function Play(){
	if(!play)
	play = setInterval( function()
		{
			//Set the cycles & Remove the players
			context.clearRect(0,0,FIELD_W+2*OFFSET_W,FIELD_H+2*OFFSET_H);

			for(var i = 0; i < 22+2; i++)
				MakeMove(content[count++]);
		}, 100);
}

function End()
{
	Pause();
	count = content.length-25*2;
	RPlayOne();
}

function Restart()
{
	Pause();
	count = 1;
	PlayOne();
}

function PrevGoal()
{

    for(var i = count; i > 1; i--)
	if(content[i].split(" ").length > 1)
	    if(content[i].split(" ")[2] == "goal_l" || content[i].split(" ")[2] == "goal_r")
	    {
		count = i-24*(CBG+48);	//Accomodate for the very large wait after a goal
		Pause();
		RPlayOne();
		return;
	    }

    window.status = "No Previous goals found.";
}

function NextGoal()
{
    for(var i = count + 24*CBG; i < content.length; i++)
    {
	if (content[i].split(" ").length > 1)
	{
	    if(content[i].split(" ")[2] == "goal_l" || content[i].split(" ")[2] == "goal_r")
	    {
		count = i-24*CBG;
		Pause();
		PlayOne();
		return;
	    }
	}
    }

    window.status = "No next goals found.";
}


function Start(file)
{
	// Check for the various File API support.
	if ( !(window.File && window.FileReader && window.FileList && window.Blob))
	{
  		alert('The File APIs are not fully supported in this browser.');
  		return false;
  	}

  	var field = document.getElementById("field");
  	context = field.getContext("2d");
  	context.font = "10pt sans-serif";
  	context.textAlign = "left";
	//var fileString = "output2.txt";

	//Run the Game
	content = $.ajax(
	{
		url: file,
		dataType: "html",
		type: "post",
		async: false,
		success: function(data) { console.log("Data Received.") },
		error: function(jqXHR, textStatus, errorThrown) { console.log("ERROR: "+errorThrown); alert("Failed to load '" + file +"'!"); }
	}).responseText.split("\n");

	//Get the team info (assumed to be in first line)
	Lteam = content[0].split(" ")[1];
	Rteam = content[0].split(" ")[2];
	Lteam = Lteam.slice(1,-1);
	Rteam = Rteam.slice(1,-1);

        L_GOALS = 0;
        R_GOALS = 0;

	document.getElementById("Lteam").innerHTML = Lteam + "&nbsp;&nbsp;" + L_GOALS;
	document.getElementById("Rteam").innerHTML = Rteam + "&nbsp;&nbsp;" + R_GOALS;	
}


function MakeMove(fileString)
{
	//Split the string into commands
	var line = fileString.split(" ");

	/*
		line[0] = T | S | b | l | L | r | R
		line[1] = stNum | Bl_x | PL_TM
		line[2] = StNam | Bl_y | PL_NUM
		line[3] =		|	   | PL_X
		line[4] =       |      | PL_Y
		line[5] =              | PL_STAMINA
	*/


	if(line[0] == "S")
	{
		$("#cyclesSlider").slider("value", line[1]);
		document.getElementById("cycles").innerHTML = line[2] + "&nbsp;&nbsp;" + line[1];

		if (L_GOALS == line[3] && R_GOALS == line[4])
		{
		    timeout = false;
		}
		else
		{
		    L_GOALS = line[3];
		    R_GOALS = line[4];
		    document.getElementById("Lteam").innerHTML = Lteam + "&nbsp;&nbsp;" + L_GOALS;
		    document.getElementById("Rteam").innerHTML = Rteam + "&nbsp;&nbsp;" + R_GOALS;
		    timeout = true;
		}
	}
	else if(line[0] == "b")
	{
		context.fillStyle = B_COLOUR;
		context.beginPath();
		context.arc( OFFSET_W+line[1]*SCALING+(FIELD_W/2), OFFSET_H+line[2]*SCALING+(FIELD_H/2) ,PLAYER_SIZE-2,0,Math.PI*2,true );
		context.closePath();
		context.fill();
	}
	else if(line[0] == "l" || line[0] == "r" || line[0] == "L" || line[0] == "R" )
	{

		if( line[0] == "R")
		    context.fillStyle = RG_COLOUR;
		else if(line[0] == "r")
		    context.fillStyle = R_COLOUR;
		else if(line[0] == "L")
		    context.fillStyle = LG_COLOUR;
		else if(line[0] == "l")
		    context.fillStyle = L_COLOUR;

		context.beginPath();
		context.arc( OFFSET_W+line[2]*SCALING+(FIELD_W/2), OFFSET_H+line[3]*SCALING+(FIELD_H/2) ,PLAYER_SIZE,0,Math.PI*2,true );
		context.closePath();
		context.fill();

		var ang0 = parseFloat(line[4]);
		var ang1 = parseFloat(line[5]);
		var x0 = OFFSET_W+line[2]*SCALING+(FIELD_W/2);
		var y0 = OFFSET_H+line[3]*SCALING+(FIELD_H/2);
		var x1 = x0 + PLAYER_SIZE*Math.cos(toRad(ang1));
		var y1 = y0 + PLAYER_SIZE*Math.sin(toRad(ang1));

		context.fillStyle = toGreyRGB(line[6]);
		context.beginPath();
		context.arc( x0, y0, PLAYER_SIZE, toRad(-90.0+ang0), toRad(90.0+ang0), true);		
		context.closePath();
		context.fill();

		context.strokeStyle = "#000000";
		context.beginPath();
		context.moveTo(x0,y0);
		context.lineTo(x1,y1);
		context.stroke();	      

		context.fillStyle = "#FFFFFF";
		context.fillText( line[1], OFFSET_W+line[2]*SCALING+(FIELD_W/2)+5, OFFSET_H+line[3]*SCALING+(FIELD_H/2) );
	}
	else if (line[0] == "T")
	{
	}
	else
	{
		console.log("ERROR: (" + line[0] + ") " + fileString);
		Pause();
	}
}


$(function(){
	var currentValue = $('#cycles');
	$("#cyclesSlider").slider({
		max: 6000,
		    min: 0,
		    slide: function(event, ui) {
		    Pause();
		    count = ui.value*24;
		    PlayOne();

		    // currentValue.html(ui.value +"/"+ 6000);
		}
	    });

    });


play.php
<!-- 
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-->
<html>
<head>
	<title>Robocup Online Player</title>
	<script language="JavaScript" src="scripts/jquery.js"></script>
	<script language="JavaScript" src="scripts/jquery-ui.js"></script>
	<script language="JavaScript" src="scripts/player.js"></script>
	<link rel="stylesheet" type="text/css" href="jquery-ui.css" />
	<link rel="stylesheet" type="text/css" href="logplay.css" />
</head>
<body>

	<div id="Notes">
		HTML5 Canvas is Required. You should get an error if it's not. <br />
	</div>

	<div id="Loader">
		Replay to load: 
		<select onchange="Start(this.value); Play();" id="fileSelect">
			<option value="">Please Select a game to play</option>
                        <?php
                                $handle = opendir('./replays');
                                while (false !== ($entry = readdir($handle)))
                                        if (substr($entry,-7) == ".replay")
                                                echo "<option value=\"./replays/". $entry ."\">". $entry ."</option>\n";
                        ?>
		</select>
	</div>

	<div id="controls">
		<div id="buttons" class="ui-widget ui-helper-clearfix">
			<div class="ui-state-default ui-corner-all" onclick="Restart();"><span class="ui-icon ui-icon-seek-first"></span></div>
			<div class="ui-state-default ui-corner-all" onclick="PrevGoal();"><span class="ui-icon ui-icon-arrowreturnthick-1-w"></span></div>
			<div class="ui-state-default ui-corner-all" onclick="RPlayOne();"><span class="ui-icon ui-icon-seek-prev"></span></div>
			<div class="ui-state-default ui-corner-all" onclick="Play();"><span class="ui-icon ui-icon-play"></span></div>
			<div class="ui-state-default ui-corner-all" onclick="Pause();"><span class="ui-icon ui-icon-pause"></span></div>
			<div class="ui-state-default ui-corner-all" onclick="PlayOne();"><span class="ui-icon ui-icon-seek-next"></span></div>
			<div class="ui-state-default ui-corner-all" onclick="NextGoal();"><span class="ui-icon ui-icon-arrowreturnthick-1-e"></span></div>
			<div class="ui-state-default ui-corner-all" onclick="End();"><span class="ui-icon ui-icon-seek-end"></span></div>
		</div>
		<div id="cyclesSlider"></div>
	</div>

	<div id="game">
		<div id="Lteam">LEFT&nbsp;&nbsp;0</div>
		<div id="cycles">0 </div>
		<div id="Rteam">RIGHT&nbsp;&nbsp;0</div>
	
		<canvas id="field" width="800" height="514">Your browser doesn't support HTML5 canvas.</canvas>
    </div>
</body>
</html>
